### DISCUSSION TITLE
递归贪心、分治

### DISCUSSION THREAD

**Discussion by TUStarry** [CSGrandeur/s-1problem1day1ac#269] [Algorithm]
# 递归

## 1. 递归定义

递归，就是在运行的过程中调用自己。

构成递归需具备的条件：

1. 子问题须与原始问题为同样的事，且更为简单；

2. 不能无限制地调用本身，须有个出口，化简为非递归状况处理。

以上是严谨的定义，有许多人给出过它的解释，但让我印象最深刻是**刘汝佳**的解释。

- **递归：**
  - **参见“递归”。**

什么？这个定义什么也没说啊！好吧，改一下：

- **递归：**
  - **如果还是没有明白什么意思，参见“递归”。**

如果你看到了现在这句话，说明你已经从上面的“递归”中出来了，而出口就是你理解了什么是“递归”。

它的特点就是自身调用自身。

## 2. 递归模板

递归要调用自己，则在代码里它要有自己（入口）；递归有出口，则会有一个return。

```c
void find(参数1){
	if(/*终止条件*/){
    	return 
    }
	find(参数2);
}
```

让我们看看下面这个蠢东西

```c
void find(参数1){
	find(参数2);    
	if(/*终止条件*/){
    	return 
    }
}
```

它有出口吗？有！但是它能走到那里吗？不行啊！（你别告诉我可以。大哥，程序是满足顺序执行的）。

实际上递归可能调用自己不止一次，并且很多递归在调用之前或调用之后都会有一些逻辑上的处理，比如下面这样。

```c
void find(参数1) {
    if (/*终止条件*/) {
        return;
    }
    //可能有一些逻辑运算
    find(参数1)
    //可能有一些逻辑运算
    find(参数2)
    //        ……
    //recursion(参数n)
    可能有一些逻辑运算
}
```

## 实战演示

> 用递归写阶乘和,int就行

```c
int rec(int n){
    if(n == 1) return 1;
    else return n * rec(n - 1);
}
```

![递归1](https://user-images.githubusercontent.com/76249637/129709853-8baabd6b-4e2e-40bf-9383-db4b47338cd5.png)

> 用递归写斐波那契数列

```C
int fib(int n){
    if(n==1) return 1;
    if(n==2) return 1;
    return fib(n - 1) + fib(n - 2);
}
```

> 二叉树的三种遍历

```c
//先序遍历
void preOrder(TreeNode node) {
    if (node == null) return;
    printf("%d ",node.val);
    preOrder(node.left);
    preOrder(node.right);
}
//中序遍历
void inOrder(TreeNode node) {
    if (node == null) return;
    inOrder(node.left);
    printf("%d ",node.val);
    inOrder(node.right);
}
//后序遍历
void postOrder(TreeNode node) {
    if (node == null) return;
    postOrder(node.left);
    postOrder(node.right);
    printf("%d ",node.val);
}
```

**多说无益，做题去。**

递归是许多算法的“元件”，像DFS就与递归息息相关，于是就选了DFS的两道题

# 分治

​		所谓分治就是指的分而治之，即将较大规模的问题分解成几个较小规模的问题，通过对较小规模问题的求解达到对整个问题的求解。当我们将问题分解成两个较小问题求解时的分治方法称之为二分法。

　　分治的基本思想是将一个规模为n的问题分解为k个规模较小的子问题，这些子问题互相独立且与原问题**相似**。找出各部分的解，然后把各部分的解组合成整个问题的解。

1. 解决算法实现的同时，需要估算算法实现所需时间。分治算法时间是这样确定的：

   解决子问题所需的工作总量（由子问题的个数、解决每个子问题的工作量 决定）合并所有子问题所需的工作量。

2. 分治法是把任意大小问题尽可能地等分成两个子问题的递归算法。

3. 分治的具体过程： 

   ```
    　 {｛开始｝
    　　 if ①问题不可分 ②返回问题解
    　　　 else {
    　　　　　 　③从原问题中划出含一半运算对象的子问题1；
    　　　　　 　　④递归调用分治法过程，求出解1；
    　　　　　 　　⑤从原问题中划出含另一半运算对象的子问题2；
    　　　　　 　　⑥递归调用分治法过程，求出解2；
    　　　　　 　　⑦将解1、解2组合成整个问题的解；
    　　　　　}
    　 } //结束
   ```

## 实战演示

> 快速排序（递归）

```c
void Quicksort(int *a, int left, int right) 
{// left、right左闭右开，low、high闭区间
    if (left >= right -1) return;
    int low = left, high = right-1, center = a[low]; 
    while(low< high)
    {
        while(low < high && a[high] >= center) high --; 
        	a[low]= a[high];
        while(low < high && a[low] <= center) low ++;
        	a[high]=a[low];        
    }
    a[low]= center;
    Quicksort(a,left, low);
    Quicksort(a,low +1,right);
}

```

> 用递归算法实现二分查找即：有n个已经从小到大排序好的数据，输入一个数m，用二分查找算法，判断它是否在这n个数中。

```c
#include<iostream>
using namespace std;
int jc(int,int);
int n,a[1000],m;
int main()
{
    int x,y,i;  
    cin>>n;
    x=1;y=n;
    for (i=1;i<=n;i++)                        //输入排序好的数
    	cin>>a[i];
    cin>>m;                                      //输入要查找的数
    jc(x,y);                                        //递归过程
    cout<<endl;    
}
int jc(int x,int y)                           //递归过程
{
    int k;
   	k=(x+y)/2;                                  //取中间位置点
 	if (a[k]==m)
		cout<<"then num in  "<<k<<endl;         //找到查找的数，输出结果
   	if (x>y) cout<<"no find"<<endl;          //找不到该数
    else
    {
       	if (a[k]<m) jc(k+1,y);                //在后半中查找
        if (a[k]>m) jc(x,k-1);                 //在前半中查找
    }
}

```

## 贪心

若在求解一个问题时，能根据每次所得到的局部最优解，推导出全局最优或最优目标。

那么，我们可以根据这个策略，每次得到局部最优解答，逐步而推导出问题，这种策略称为贪心法。

狭义的贪心算法指的是解最优化问题的一种特殊方法，解决过程中总是做出当下最好的选择，因为具有最优子结构的特点，局部最优解可以得到全局最优解；这种贪心算法是动态规划的一种特例。能用贪心解决的问题，也可以用动态规划解决。

而广义的贪心指的是一种通用的贪心策略，基于当前局面而进行贪心决策。

**适用条件**

1. 符合贪心策略

   所谓贪心选择性质是指所求问题的整体最优解可以通过一系列局部最优的选择，即贪心选择来达到。这是贪心算法可行的第一个基本要素，也是贪心算法与动态规划算法的主要区别。

   贪心算法则通常以**自顶向下的方式进行**，以迭代的方式作出相继的贪心选择，每作一次贪心选择就将所求问题简化为规模更小的子问题。

   对于一个具体问题，要确定它是否具有贪心选择性质，必须证明每一步所作的贪心选择最终导致问题的整体最优解。

2. 最优子结构性质

   当一个问题的最优解包含其子问题的最优解时，称此问题具有最优子结构性质。问题的最优子结构性质是该问题可用动态规划算法或贪心算法求解的关键特征。

![递归2](https://user-images.githubusercontent.com/76249637/129709882-9ca279c5-29fb-4f6a-9c07-3bd9de4766ea.png)

## 实战演练

> 在N行M列的正整数矩阵中，要求从每行中选出1个数，使得选出的总共N个数的和最大。

```c
Total = 0;
for (int I = 1;  I <= N; ++ l) 
{	//对N行进行选择
    选择第I行最大的数,记为K；
    Total += K；
}
```

> 给定一个最大载重量为M的卡车和N种食品，有食盐，白糖，大米等。已知第i种食品的最多拥有Wi公斤，其商品价值为Vi元/公斤，编程确定一个装货方案，使得装入卡车中的所有物品总价值最大。

​		因为每一个物品都可以分割成单位块，单位块的利益越大显然总收益越大，所以它局部最优满足全局最优，可以用贪心法解答，方法如下：先将单位块收益按从大到小进行排列，然后用循环从单位块收益最大的取起，直到不能取为止便得到了最优解。

```c
i=1;
do
{  
　　if (m==0)  break; 	                   //如果卡车满载则跳出循环
　　m=m-w[i];
　　if (m>=0)                                  //将第i种商品全部装入卡车
　　else  将(m+w[i])  重量的物品i装入卡车;
　　i=i+1;             	                   //选择下一种商品
}while (m>0&&i<=n);
```

**Comment by TUStarry** [comment]
HDU1979(BFS解法):https://paste.ubuntu.com/p/zPXdCT86py/

**Reply by TUStarry** [reply]
HDU1979(DFS解法):https://paste.ubuntu.com/p/FkgRnKPZ9P/

**Reply by CSGrandeur** [reply]
放你的题解仓库不是更好
